File: apps/runtime/src/git/diff.rs
//! Working-tree and index diffs: `git diff` and `git diff --cached`.

use super::*;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitFileDiff {
    pub path: String,
    /// Normalized to `M` / `A` / `D` / `R` / `?`.
    pub status: String,
    #[serde(skip)]
    pub untracked: bool,
    pub additions: u64,
    pub deletions: u64,
    pub patch: String,
    /// `false` when the file is listed but its content cannot be shown as a
    /// textual patch (binary or oversized untracked file). `patch` is then
    /// empty and must not be exported as part of a unified diff.
    pub previewable: bool,
    /// The patch comes from the index (`git diff --cached`) rather than the
    /// working tree. Mirrors the requested scope.
    pub staged: bool,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitDiff {
    pub repository: bool,
    pub clean: bool,
    pub files: Vec<GitFileDiff>,
}

/// Which side of the index a diff is taken from (plan §3.6).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum DiffScope {
    /// Unstaged edits plus untracked files — what `git diff` shows.
    #[default]
    Worktree,
    /// What `git diff --cached` shows. Untracked files never appear here.
    Staged,
}

/// Request shape for [`read_diff`].
#[derive(Debug, Clone, Default)]
pub struct DiffRequest {
    pub scope: DiffScope,
    /// Restricts the diff to these workspace-relative paths. Empty means "use
    /// the requested directory as the pathspec".
    pub paths: Vec<String>,
    /// `--ignore-all-space`: hides whitespace-only differences from the patch
    /// and from the line counts. This is a display option — nothing is staged,
    /// applied, or committed from a whitespace-ignoring diff.
    pub ignore_whitespace: bool,
}

/// `GET /api/workspaces/{id}/git/diff?scope=&paths=`.
///
/// `scope` picks the side of the index: `worktree` is `git diff` plus untracked
/// files, `staged` is `git diff --cached` and never lists untracked files
/// (there is nothing staged about them).
/// Trusted internal entry point: configured helpers may run. Workspace-scoped
/// callers must pass their grant to `read_diff_with_execution` instead.
pub fn read_diff(
    workspace_root: &Path,
    requested: &str,
    request: &DiffRequest,
) -> AppResult<GitDiff> {
    read_diff_with_execution(workspace_root, requested, request, true)
}
pub fn read_diff_with_execution(
    workspace_root: &Path,
    requested: &str,
    request: &DiffRequest,
    execute: bool,
) -> AppResult<GitDiff> {
    if request.scope == DiffScope::Worktree {
        access::require_execution(execute, "Git worktree diff")?;
    }
    let Some(context) = repo_context_with_execution(workspace_root, requested, execute)? else {
        return Ok(GitDiff {
            repository: false,
            clean: true,
            files: vec![],
        });
    };
    let RepoContext {
        workspace_root,
        repository,
        pathspec,
    } = context;

    let pathspecs = if request.paths.is_empty() {
        if pathspec.is_empty() {
            vec![]
        } else {
            vec![pathspec]
        }
    } else {
        if request.paths.len() > MAX_PATHS_PER_REQUEST {
            return Err(AppError::BadRequest(
                "At most 200 paths can be diffed at once".into(),
            ));
        }
        request
            .paths
            .iter()
            .map(|path| workspace_relative_path(path))
            .collect::<AppResult<Vec<_>>>()?
    };

    let cached = request.scope == DiffScope::Staged;
    let mut files = BTreeMap::<String, GitFileDiff>::new();
    for (status, path) in diff_name_status(&repository, cached, &pathspecs, execute)? {
        files.insert(
            path.clone(),
            GitFileDiff {
                path,
                status,
                untracked: false,
                additions: 0,
                deletions: 0,
                patch: String::new(),
                previewable: true,
                staged: cached,
            },
        );
    }
    if !cached {
        for entry in status_entries(&repository, &pathspecs)? {
            if entry.status == "?" && !files.contains_key(&entry.path) {
                files.insert(
                    entry.path.clone(),
                    GitFileDiff {
                        path: entry.path,
                        status: "?".to_owned(),
                        untracked: true,
                        additions: 0,
                        deletions: 0,
                        patch: String::new(),
                        previewable: true,
                        staged: false,
                    },
                );
            }
        }
    }

    // The file list itself never ignores whitespace: a whitespace-only edit is
    // still a change, and hiding the row would contradict `git status`. Only
    // the rendered patch and its line counts honour the option, so such a file
    // shows up with an empty patch and 0/0 — which is exactly what it is.
    for (additions, deletions, path) in diff_numstat(
        &repository,
        cached,
        &pathspecs,
        execute,
        request.ignore_whitespace,
    )? {
        if let Some(file) = files.get_mut(&path) {
            file.additions += additions;
            file.deletions += deletions;
        }
    }

    for file in files.values_mut() {
        if !file.untracked {
            let path = file.path.clone();
            let mut args = vec!["diff"];
            if cached {
                args.push("--cached");
            }
            if request.ignore_whitespace {
                args.push(IGNORE_WHITESPACE);
            }
            args.extend(["--", path.as_str()]);
            file.patch = git_with_execution(&repository, &args, execute)?
                .trim_end_matches('\n')
                .to_owned();
        } else {
            // A single un-previewable untracked file (binary or oversized)
            // must not fail the whole scan; it is listed without a textual
            // patch instead.
            match read_untracked_file(&workspace_root, &repository, &file.path) {
                Ok(content) => {
                    file.additions = content.lines().count() as u64;
                    file.patch = content
                        .lines()
                        .take(400)
                        .map(|line| format!("+{line}"))
                        .collect::<Vec<_>>()
                        .join("\n");
                }
                // Symlinks / paths escaping the workspace stay hard errors.
                Err(AppError::BadRequest(reason)) => {
                    tracing::debug!(path = %file.path, %reason, "untracked file not previewable");
                    file.additions = 0;
                    file.patch.clear();
                    file.previewable = false;
                }
                Err(error) => return Err(error),
            }
        }
    }

    Ok(GitDiff {
        repository: true,
        clean: files.is_empty(),
        files: files.into_values().collect(),
    })
}

/// `git diff [--cached] --name-status -z` → `(status, path)`.
///
/// `-z` is what makes this safe: paths are emitted verbatim, so nothing has to
/// be un-quoted and a path containing a space, quote or newline still parses.
/// Renames and copies emit `R100\0<old>\0<new>\0`; the new path is the one the
/// UI addresses.
fn diff_name_status(
    repository: &Path,
    cached: bool,
    pathspecs: &[String],
    execute: bool,
) -> AppResult<Vec<(String, String)>> {
    let output = git_with_execution(
        repository,
        &diff_args("--name-status", cached, pathspecs),
        execute,
    )?;
    let mut fields = output.split('\0').filter(|field| !field.is_empty());
    let mut entries = Vec::new();
    while let Some(code) = fields.next() {
        let renamed = code.starts_with('R') || code.starts_with('C');
        let Some(first) = fields.next() else { break };
        let path = if renamed {
            let Some(second) = fields.next() else { break };
            second
        } else {
            first
        };
        entries.push((normalize_file_status(code), path.to_owned()));
    }
    Ok(entries)
}

/// `git diff [--cached] --numstat -z` → `(additions, deletions, path)`.
///
/// Binary files are reported as `-\t-\t<path>`, which parses to `0 / 0`.
/// Renames put the paths in their own NUL fields and leave the inline path
/// empty (`1\t2\t\0<old>\0<new>\0`).
fn diff_numstat(
    repository: &Path,
    cached: bool,
    pathspecs: &[String],
    execute: bool,
    ignore_whitespace: bool,
) -> AppResult<Vec<(u64, u64, String)>> {
    let mut args = diff_args("--numstat", cached, pathspecs);
    if ignore_whitespace {
        // Before the `--` separator, so it is read as an option and not as a
        // pathspec.
        args.insert(if cached { 2 } else { 1 }, IGNORE_WHITESPACE);
    }
    let output = git_with_execution(repository, &args, execute)?;
    let mut fields = output.split('\0').filter(|field| !field.is_empty());
    let mut entries = Vec::new();
    while let Some(record) = fields.next() {
        let mut parts = record.splitn(3, '\t');
        let additions = parts
            .next()
            .and_then(|value| value.parse().ok())
            .unwrap_or(0);
        let deletions = parts
            .next()
            .and_then(|value| value.parse().ok())
            .unwrap_or(0);
        let inline = parts.next().unwrap_or_default();
        let path = if inline.is_empty() {
            let _origin = fields.next();
            match fields.next() {
                Some(destination) => destination.to_owned(),
                None => break,
            }
        } else {
            inline.to_owned()
        };
        entries.push((additions, deletions, path));
    }
    Ok(entries)
}

fn diff_args<'a>(mode: &'a str, cached: bool, pathspecs: &'a [String]) -> Vec<&'a str> {
    let mut args = vec!["diff"];
    if cached {
        args.push("--cached");
    }
    args.extend([mode, "-z", "--"]);
    args.extend(pathspecs.iter().map(String::as_str));
    args
}

fn read_untracked_file(workspace_root: &Path, repository: &Path, path: &str) -> AppResult<String> {
    const MAX_UNTRACKED_PREVIEW: u64 = 1024 * 1024;
    let candidate = repository.join(path);
    let metadata = std::fs::symlink_metadata(&candidate)
        .map_err(|_| AppError::BadRequest("Untracked file cannot be inspected".into()))?;
    if metadata.file_type().is_symlink() || !metadata.is_file() {
        return Err(AppError::Forbidden(
            "Untracked symlinks and special files are not previewed".into(),
        ));
    }
    if metadata.len() > MAX_UNTRACKED_PREVIEW {
        return Err(AppError::BadRequest(
            "Untracked file is too large to preview".into(),
        ));
    }
    let canonical = candidate
        .canonicalize()
        .map_err(|_| AppError::BadRequest("Untracked file cannot be resolved".into()))?;
    if !canonical.starts_with(workspace_root) || !canonical.starts_with(repository) {
        return Err(AppError::Forbidden(
            "Untracked file is outside the authorized workspace".into(),
        ));
    }
    let mut bytes = Vec::with_capacity(metadata.len() as usize);
    File::open(&canonical)?
        .take(MAX_UNTRACKED_PREVIEW)
        .read_to_end(&mut bytes)?;
    String::from_utf8(bytes)
        .map_err(|_| AppError::BadRequest("Untracked binary files are not previewed".into()))
}

File: apps/desktop/src/core/git/diff.ts
import { openSync, readSync, closeSync, lstatSync } from "node:fs";
import { join } from "node:path";
import { canonicalize, workspaceRelativePath } from "../workspaces/roots";
import {
  IGNORE_WHITESPACE,
  MAX_PATHS_PER_REQUEST,
  gitText,
  isInside,
  repoContext,
} from "./context";
import { normalizeFileStatus, statusEntries } from "./status";
import { badRequest, forbidden, requireExecution } from "./support";
import { DomainError } from "../workspaces/support";

/**
 * Working-tree and index diffs: `git diff` and `git diff --cached`.
 *
 * A port of `apps/runtime/src/git/diff.rs`.
 */

export type DiffScope = "worktree" | "staged";

export interface GitFileDiff {
  readonly path: string;
  readonly status: string;
  readonly additions: number;
  readonly deletions: number;
  readonly patch: string;
  /**
   * `false` when the file is listed but its content cannot be shown as a
   * textual patch (binary, or an oversized untracked file).
   */
  readonly previewable: boolean;
  readonly staged: boolean;
}

export interface GitDiff {
  readonly repository: boolean;
  readonly clean: boolean;
  readonly files: GitFileDiff[];
}

export interface DiffRequest {
  readonly scope: DiffScope;
  readonly paths: readonly string[];
  readonly ignoreWhitespace: boolean;
}

interface MutableFileDiff {
  path: string;
  status: string;
  untracked: boolean;
  additions: number;
  deletions: number;
  patch: string;
  previewable: boolean;
  staged: boolean;
}

const MAX_UNTRACKED_PREVIEW = 1024 * 1024;

export async function readDiff(
  workspaceRoot: string,
  requested: string,
  request: DiffRequest,
  execute = true,
): Promise<GitDiff> {
  if (request.scope === "worktree") {
    requireExecution(execute, "Git worktree diff");
  }
  const context = await repoContext(workspaceRoot, requested, execute);
  if (context === undefined) {
    return { repository: false, clean: true, files: [] };
  }
  const { workspaceRoot: root, repository, pathspec } = context;

  let pathspecs: string[];
  if (request.paths.length === 0) {
    pathspecs = pathspec === "" ? [] : [pathspec];
  } else {
    if (request.paths.length > MAX_PATHS_PER_REQUEST) {
      throw badRequest("At most 200 paths can be diffed at once");
    }
    pathspecs = request.paths.map((path) => workspaceRelativePath(path));
  }

  const cached = request.scope === "staged";
  const files = new Map<string, MutableFileDiff>();
  for (const [status, path] of await diffNameStatus(
    repository,
    cached,
    pathspecs,
    execute,
  )) {
    files.set(path, {
      path,
      status,
      untracked: false,
      additions: 0,
      deletions: 0,
      patch: "",
      previewable: true,
      staged: cached,
    });
  }
  if (!cached) {
    for (const entry of await statusEntries(repository, pathspecs)) {
      if (entry.status === "?" && !files.has(entry.path)) {
        files.set(entry.path, {
          path: entry.path,
          status: "?",
          untracked: true,
          additions: 0,
          deletions: 0,
          patch: "",
          previewable: true,
          staged: false,
        });
      }
    }
  }

  // The file list itself never ignores whitespace: a whitespace-only edit is
  // still a change, and hiding the row would contradict `git status`. Only the
  // rendered patch and its line counts honour the option.
  for (const [additions, deletions, path] of await diffNumstat(
    repository,
    cached,
    pathspecs,
    execute,
    request.ignoreWhitespace,
  )) {
    const file = files.get(path);
    if (file !== undefined) {
      file.additions += additions;
      file.deletions += deletions;
    }
  }

  for (const file of files.values()) {
    if (!file.untracked) {
      const args = ["diff"];
      if (cached) args.push("--cached");
      if (request.ignoreWhitespace) args.push(IGNORE_WHITESPACE);
      args.push("--", file.path);
      file.patch = (await gitText(repository, args, execute)).replace(
        /\n+$/,
        "",
      );
      continue;
    }
    // A single un-previewable untracked file must not fail the whole scan; it
    // is listed without a textual patch instead.
    try {
      const content = readUntrackedFile(root, repository, file.path);
      const lines = content.split("\n");
      const trailing = lines.length > 0 && lines[lines.length - 1] === "";
      const counted = trailing ? lines.slice(0, -1) : lines;
      file.additions = counted.length;
      file.patch = counted
        .slice(0, 400)
        .map((line) => `+${line}`)
        .join("\n");
    } catch (error) {
      if (error instanceof DomainError && error.status === 400) {
        file.additions = 0;
        file.patch = "";
        file.previewable = false;
        continue;
      }
      throw error;
    }
  }

  const ordered = [...files.keys()].sort().map((path) => {
    const file = files.get(path) as MutableFileDiff;
    return {
      path: file.path,
      status: file.status,
      additions: file.additions,
      deletions: file.deletions,
      patch: file.patch,
      previewable: file.previewable,
      staged: file.staged,
    };
  });
  return { repository: true, clean: ordered.length === 0, files: ordered };
}

function diffArgs(
  mode: string,
  cached: boolean,
  pathspecs: readonly string[],
): string[] {
  const args = ["diff"];
  if (cached) args.push("--cached");
  args.push(mode, "-z", "--", ...pathspecs);
  return args;
}

/**
 * `git diff [--cached] --name-status -z` → `(status, path)`.
 *
 * `-z` is what makes this safe: paths are emitted verbatim, so a path
 * containing a space, quote or newline still parses. Renames and copies emit
 * `R100\0<old>\0<new>\0`; the new path is the one the UI addresses.
 */
async function diffNameStatus(
  repository: string,
  cached: boolean,
  pathspecs: readonly string[],
  execute: boolean,
): Promise<[string, string][]> {
  const output = await gitText(
    repository,
    diffArgs("--name-status", cached, pathspecs),
    execute,
  );
  const fields = output.split("\0").filter((field) => field !== "");
  const entries: [string, string][] = [];
  for (let index = 0; index < fields.length; ) {
    const code = fields[index] as string;
    index += 1;
    const renamed = code.startsWith("R") || code.startsWith("C");
    const first = fields[index];
    if (first === undefined) break;
    index += 1;
    let path = first;
    if (renamed) {
      const second = fields[index];
      if (second === undefined) break;
      index += 1;
      path = second;
    }
    entries.push([normalizeFileStatus(code), path]);
  }
  return entries;
}

/**
 * `git diff [--cached] --numstat -z` → `(additions, deletions, path)`.
 *
 * Binary files are reported as `-\t-\t<path>`, which parses to `0 / 0`.
 * Renames put the paths in their own NUL fields and leave the inline path
 * empty.
 */
async function diffNumstat(
  repository: string,
  cached: boolean,
  pathspecs: readonly string[],
  execute: boolean,
  ignoreWhitespace: boolean,
): Promise<[number, number, string][]> {
  const args = diffArgs("--numstat", cached, pathspecs);
  if (ignoreWhitespace) {
    // Before the `--` separator, so it is read as an option and not a pathspec.
    args.splice(cached ? 2 : 1, 0, IGNORE_WHITESPACE);
  }
  const output = await gitText(repository, args, execute);
  const fields = output.split("\0").filter((field) => field !== "");
  const entries: [number, number, string][] = [];
  for (let index = 0; index < fields.length; ) {
    const record = fields[index] as string;
    index += 1;
    const firstTab = record.indexOf("\t");
    const secondTab = record.indexOf("\t", firstTab + 1);
    const additions = Number.parseInt(record.slice(0, firstTab), 10);
    const deletions = Number.parseInt(
      record.slice(firstTab + 1, secondTab),
      10,
    );
    const inline = secondTab < 0 ? "" : record.slice(secondTab + 1);
    let path: string;
    if (inline === "") {
      index += 1; // the origin path
      const destination = fields[index];
      if (destination === undefined) break;
      index += 1;
      path = destination;
    } else {
      path = inline;
    }
    entries.push([
      Number.isNaN(additions) ? 0 : additions,
      Number.isNaN(deletions) ? 0 : deletions,
      path,
    ]);
  }
  return entries;
}

function readUntrackedFile(
  workspaceRoot: string,
  repository: string,
  path: string,
): string {
  const candidate = join(repository, path);
  const info = lstatSync(candidate, { throwIfNoEntry: false });
  if (info === undefined) {
    throw badRequest("Untracked file cannot be inspected");
  }
  if (info.isSymbolicLink() || !info.isFile()) {
    throw forbidden("Untracked symlinks and special files are not previewed");
  }
  if (info.size > MAX_UNTRACKED_PREVIEW) {
    throw badRequest("Untracked file is too large to preview");
  }
  let canonical: string;
  try {
    canonical = canonicalize(candidate);
  } catch {
    throw badRequest("Untracked file cannot be resolved");
  }
  if (!isInside(workspaceRoot, canonical) || !isInside(repository, canonical)) {
    throw forbidden("Untracked file is outside the authorized workspace");
  }
  const handle = openSync(canonical, "r");
  try {
    const buffer = Buffer.alloc(Math.min(info.size, MAX_UNTRACKED_PREVIEW));
    const read = readSync(handle, buffer, 0, buffer.byteLength, 0);
    const bytes = buffer.subarray(0, read);
    const text = bytes.toString("utf8");
    if (Buffer.from(text, "utf8").compare(bytes) !== 0) {
      throw badRequest("Untracked binary files are not previewed");
    }
    return text;
  } finally {
    closeSync(handle);
  }
}
